typing/typedecl_separability.mli

```diff
@@ -0,0 +1,125 @@
+
+(** The OCaml runtime assumes for type-directed optimizations that all types
+    are "separable". A type is "separable" if either all its inhabitants
+    (the values of this type) are floating-point numbers, or none of them are.
+
+    (Note: This assumption is required for the dynamic float array optimization;
+    it is only made if Config.flat_float_array is set,
+    otherwise the code in this module defaults becomes trivial
+    -- see {!compute_decl}.)
+
+    This soundness requirement could be broken by type declarations mixing
+    existentials and the "[@@unboxed]" annotation. Consider the declaration
+
+    {[
+       type any = Any : 'a -> any [@@unboxed]
+    ]}
+
+   which corresponds to the existential type "exists a. a". If this type is
+   allowed to be unboxed, then it is inhabited by both [float] values
+   and non-[float] values. On the contrary, if unboxing is disallowed, the
+   inhabitants are all blocks with the [Any] constructors pointing to its
+   parameter: they may point to a float, but they are not floats.
+
+   The present module contains a static analysis ensuring that declarations
+   annotated with "[@@unboxed]" can be safely unboxed. The idea is to check
+   the "separability" (in the above sense) of the argument type that would
+   be unboxed, and reject the unboxed declaration if it would create a
+   non-separable type.
+
+   Checking mutually-recursive type declarations is a bit subtle.
+   Consider, for example, the following declarations.
+
+   {[
+      type foo = Foo : 'a t -> foo   [@@unboxed]
+      and 'a t = ...
+   ]}
+
+   Deciding whether the type [foo] should be accepted requires inspecting
+   the declaration of ['a t], which may itself refer to [foo] in turn.
+   In general, the analysis performs a fixpoint computation. It is somewhat
+   similar to what is done for inferring the variance of type parameters.
+
+   Our analysis is defined using inference rules "Def; Gamma |- t : m", in
+   which a type expression "t" is checked agains a "mode" "m". This "mode"
+   describes the separability requirement on the type expression (see below
+   for more details). In the inference rules, [Gamma] maps type variables to
+   modes and [Def] records the "mode signature" of the mutually-recursive
+   type declarations that are being checked.
+
+   The "mode signature" of a type with parameters [('a, 'b) t] is of the
+   form [('a : m1, 'b : m2) t], where [m1] and [m2] are modes. Its meaning
+   is the following: a concrete instance [(foo, bar) t] of the type is
+   separable if [foo] has mode [m1] and [bar] has mode [m2]. *)
+
+type error =
+  | Non_separable_evar of string option
+exception Error of Location.t * error
+(** Exception raised when a type declaration is not separable, or when its
+    separability cannot be established. *)
+
+type mode = Types.Separability.t = Ind | Sep | Deepsep
+(** The mode [Sep] ("separable") is used on type expressions that must be
```

Comment:
The comment phrases the mode as a requirement on arguments ("type expressions that must be separable"), and that's generally how it's used.  But I'd prefer to see it defined more neutrally somewhere as a description of type expressions.  One option is to document `Types.Separability.t` as a (representation of a) property of type expressions, and separately document `Typedecl_separability.mode` as a requirement on arguments, as it is currently.

Reply:
I rewrote this comment to include for each mode, then the type-expression view, and then the type-parameter view. I think the latter is necessary to give intuition (type expressions themselves, at the toplevel, must all be separable).